Scala fast test

ittaiz opened the pull request:
Trying to follow the work by @domasmart https://github.com/wix-playground/intellij/pull/66
I've cherry-picked the commit to keep history of author and resolved conflicts.
Build is failing because there are logical issues apart from git. 
Will work on those next

ittaiz commented:
Commits and files are messed up since I rebased against V21.0 to be able to send Tzufit a working plugin

ittaiz commented:
@liucijus I think this is nearing a stable state.
I have a few more things to handle but you can take a look if you want.
I'm going on vacation and will be back in ~4 weeks.

ittaiz commented:
@liucijus @simuons
I think we're ready.
There are a few open issues which I think (and you can definitely disagree) are not blocking for merge:

- [x] FastBuildCompilerFactoryImpl- Hacky way to substitute FastBuildCompiler with Scala compiler
- [x] BazelFastBuildTestEnvironmentCreator- Find JDK
- [x] Hyperlink and console- Check if we can have clickable errors without hyperlink like java has them (BspNotificationsHandler, StatusOutput, ConsoleView,, ErrorLocation) 
- [ ] Any better solution for FileUtils recursive copy of folders?
- [ ] .sh.runfiles change- will be handled once rules_scala PR is merged
- [ ] output_groups- need to agree with upstream on strategy

Apart from that we have a few open features (Proto change, AutoDep, ITs, test filtering) and some more minor ones which will be tracked separately.

ittaiz commented:
forgot to tag @rotemmiz

ittaiz commented:
@liucijus the [JavaExe exec path taken from java target in bazel](https://github.com/wix-playground/intellij/pull/71/commits/964f7425aa182d9dd9c23a4b830fa27edd736a1c)
[964f742](https://github.com/wix-playground/intellij/pull/71/commits/964f7425aa182d9dd9c23a4b830fa27edd736a1c) commit is one I'm not 100% sure about code wise. I'm not sure if `getJavaExecutableExecPath` should sit in `FastBuildTestEnvironmentCreator` or `BazelFastBuildTestEnvironmentCreator`.
Also I made two classes in `FastBuildBlazeData` public. Not sure about that.
Apart from that it's probably ok.
Only works from bazel 5.0 but it should just require local_jdk like today for earlier versions.

ittaiz commented:
I might try to handle `Hyperlink and console` pre merge but really unsure.
I think as is we can get it into master and announce beta.

ittaiz commented:
Ok, I think that this can be merged. 
I would estimate the following might cause friction when trying to upstream:

- [ ]  Any better solution for FileUtils recursive copy of folders?
- [ ]  .sh.runfiles change- will be handled once rules_scala PR is merged
- [ ] output_groups- need to agree with upstream on strategy 
- [ ] Missing tests

1-3 are really small.
4 is bigger but I'm thinking of maybe handling in a followup PR.

liucijus commented on java/src/com/google/idea/blaze/java/fastbuild/FastBuildBlazeData.java:
```diff
@@ -224,13 +224,16 @@ abstract static class JavaToolchainInfo {
         String sourceVersion,
-        String targetVersion) {
+        String targetVersion,
+        Optional<JavaRuntime> javaRuntime) {
```
Idiomatic use of Optional type in Java is only for return types

ittaiz replied:
fixed please see new commit

ittaiz commented on java/src/com/google/idea/blaze/java/run/fastbuild/FastBuildTestEnvironmentCreator.java:
```diff
@@ -140,10 +142,27 @@ GeneralCommandLine createCommandLine(
+      // Wix tooling injects some java arguments which makes java call fail
+      // Temporary filtering them out to see if it's makes run succeed
+      // ex: "-javaagent:$(rootpath @core_server_build_tools//test-agent/src/main/java/com/wixpress/agent:test-agent_deploy.jar)"
+      if (flag.contains("@core_server_build_tools")) {
```
this is actually another hack which we should track. It's related though to IT running so maybe that's enough

liucijus reviewed (approved):
@ittaiz before we merge, can you reduce the number of commits to some logical groups, or shall we just squash?

ittaiz committed:
Fast test supports Scala junit
Via running bloop and BSP

ittaiz committed:
require the output groups
for some reason without the + the runfiles dir isn't created
this is something that native rules don't suffer from
another alternative is to add the _hidden_top_level_INTERNAL_ output group

ittaiz committed:
HACK due to rules_scala change with the binary script name

ittaiz committed:
HACK to allow Wix tests to run
Likely needs to be handled to support ITs (and to be upstreamed)

ittaiz commented:
@liucijus we're good to go.
You'll rebase and merge so later on we can drop the "hack rules_scala" commit?
